### DISCUSSION TITLE
Save current machine state

### DISCUSSION THREAD

**Discussion by kev-wes** [nasa/prog_models#255] [Q&A] [answered]
First of all, I want to thank you for the prog_models package, which is impressive! I have a question and could not find any solution after reading all the docs and tutorials.

I want to use the package for simulating a self-adjusting battery. The battery is used with intensity for a fixed period `t`, e.g., `100` (using `future_loading`). It then looks at its RUL and adjusts its intensity. However, I run into some run time issues as I always call a fresh battery and have to simulate from `t=0`. An example:
Time | EOD | Intensity
000 | 1.00 | 
Then, the battery uses intensity 4 for the following 100 time indices because the EOD is still high. I adjust the `future_loading` function and `simulate_to` `t=100`.
Time | EOD | Intensity
000 | 1.00 | 4
100 | 0.96 | 
Now, the battery reduces its intensity to 3 based on the observations (the reason is irrelevant to the issue). Now the problem arises that I have to `simulate_to` from `t=0` again. I do this by adjusting the `future_loading` function and `simulate_to` `t=200`. Here, the time span between 0 and 100 did not change, but I still have to simulate it again,  which costs time and, due to stochasticity, might lead to slightly different results (which is not that relevant to me).
Time | EOD | Intensity
000 | 1.00 | 4	<- is executed again (redundant)
100 | 0.96 | 3
200 | 0.93 | 

Is there a way to, e.g., store the whole battery object with its current conditions, including its current degradation level, and then call it again and simulate it further without starting from its initial state?

**Comment by kev-wes** [comment, marked as answer]
I solved this by myself, which, in hindsight, is quite trivial. The return values of `simulate_to` comprise a list of states. Setting these states as new `model.parameters` saves the current state of the model. 

**Example**
Simulation for 100 time steps (assume that `future_loading` always returns a specific intensity for any t, e.g., 2):
`(_, _, states, _, event_states) = batt.simulate_to(100, future_loading, **options)`
Set states of battery to final values of last observation (i.e., after simulating 100 time steps with intensity 2):
`batt.parameters['x0'] = states[-1]`
Now any following simulation starts as if the battery was already run with intensity 2 for 100 time steps.

**Reply by teubert** [reply]
Sorry, just saw this question. I need to start keeping an eye on the discussion. 

Yes, what you're describing perfectly. You can also specify the initial state in simulation using the x parameters. For example:

```python
(_, _, states, _, event_states) = batt.simulate_to(100, future_loading, **options, x = states[-1])
```

This will work more constantly as not every model uses x0 to set the state.
